### PR TITLE
Generate labels in MIX

### LINKED ISSUE #210: Classifier: get_labels of NN does not return mixed labels
``labels_`` is not mixed.
https://github.com/jubatus/jubatus_core/blob/0.2.5/jubatus/core/classifier/nearest_neighbor_classifier.hpp#L74

### PR DESCRIPTION
**Pull request by rimms**
Fixes #210

Note: this patch based on #270

### CODE REVIEW

**Comment by kmaehashi** [comment]
@rimms Rebase please.

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp [hunk 1]
```diff
@@ -256,6 +260,29 @@ void nearest_neighbor_classifier::unlearn_id(const std::string& id) {
   nearest_neighbor_engine_->get_table()->delete_row(id);
 }
 
+void nearest_neighbor_classifier::unlearn_label(const std::string& label) {
```
Comment (kmaehashi) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: The method name and its function mismatch.  How about `decrement_label_counter`?

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp [hunk 2]
```diff
@@ -256,6 +260,29 @@ void nearest_neighbor_classifier::unlearn_id(const std::string& id) {
...
+  util::concurrent::scoped_lock lk(label_mutex_);
+  labels_[label] -= 1;
+  if (labels_[label] <= 0) {
+    labels_.erase(label);
+  }
+}
+
+void nearest_neighbor_classifier::regenerate_labels() {
```
Comment (kmaehashi) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: `regenerate_label_counter`

File: jubatus/core/classifier/nearest_neighbor_classifier.hpp [hunk 3]
```diff
@@ -71,7 +71,8 @@ class nearest_neighbor_classifier : public classifier_base {
  private:
   jubatus::util::lang::shared_ptr<nearest_neighbor::nearest_neighbor_base>
       nearest_neighbor_engine_;
-  jubatus::util::data::unordered_set<std::string> labels_;
+  typedef jubatus::util::data::unordered_map<std::string, size_t> labels_t;
+  labels_t labels_;
```
Comment (kmaehashi) [inline jubatus/core/classifier/nearest_neighbor_classifier.hpp]: For complexed data structures its better to add comment to improve readability.

```
// A map from label to number of records that belongs to the label.
```

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp [hunk 4]
```diff
@@ -256,6 +260,29 @@ void nearest_neighbor_classifier::unlearn_id(const std::string& id) {
...
+  shared_ptr<const storage::column_table> table =
```
Comment (kmaehashi) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: Should we lock a table here?

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp [hunk 5]
```diff
@@ -256,6 +260,29 @@ void nearest_neighbor_classifier::unlearn_id(const std::string& id) {
...
+
+  labels_t labels;
+  for (size_t i = 0, n = table->size(); i < n; ++i) {
+    std::string id = table->get_key(i);
+    std::string label = get_label_from_id(id);
+    labels[label] += 1;
+  }
+
+  util::concurrent::scoped_lock lk(label_mutex_);
+  labels_.swap(labels);
```
Comment (kmaehashi) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: Labels added via `set_label` API will be lost here.  Copying keyset from `labels_` seems better.

**Comment by kmaehashi** [comment]
@rimms Thanks!  Added some comments.

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp (hunk 1 above)
Comment (rimms) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: OK, I will rename this function to ``decrement_label_counter`` .

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp (hunk 2 above)
Comment (rimms) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: OK, I will rename this function to ``regenerate_label_counter`` .

File: jubatus/core/classifier/nearest_neighbor_classifier.hpp (hunk 3 above)
Comment (rimms) [inline jubatus/core/classifier/nearest_neighbor_classifier.hpp]: That's right. I will add the above comment.

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp (hunk 4 above)
Comment (rimms) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: I will add the followings:
```cpp
util::concurrent::scoped_rlock table_lk(table->get_mutex());
```

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp (hunk 5 above)
Comment (rimms) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: That's right.
I didn’t consider about ``set_label`` . Thank you.

**Comment by rimms** [comment]
@kmaehashi Thanks for review.
I rebased this patch and fixed.

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp [hunk 6]
```diff
@@ -260,6 +260,41 @@ void nearest_neighbor_classifier::unlearn_id(const std::string& id) {
...
+void nearest_neighbor_classifier::decrement_label_counter(
+    const std::string& label) {
+  util::concurrent::scoped_lock lk(label_mutex_);
+  labels_[label] -= 1;
+  if (labels_[label] <= 0) {
+    labels_.erase(label);
+  }
+}
+
+void nearest_neighbor_classifier::regenerate_label_counter() {
```
Comment (kmaehashi) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: It's better to revise names of variable names; difficult to detect typos between `labels` and `labels_`

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp [hunk 7]
```diff
@@ -260,6 +260,41 @@ void nearest_neighbor_classifier::unlearn_id(const std::string& id) {
...
+      labels[label] += 1;
+    }
+  }
+
+  util::concurrent::scoped_lock lk(label_mutex_);
+  for (labels_t::const_iterator iter = labels_.begin();
+       iter != labels_.end(); ++iter) {
+    if (labels.find(iter->first) == labels.end()) {
+      // labels registered by set_label are exist
+      labels[iter->first] = iter->second;
```
Comment (kmaehashi) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: `iter->second` should be `0` instead?

**Comment by kmaehashi** [comment]
@rimms Added comments. Please catch up with inverted index classifiers (#276).

**Comment by rimms** [comment]
@kmaehashi Thanks for review. I fixed.
I will catch up #276 in #272 .

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp [hunk 8]
```diff
@@ -235,6 +235,44 @@ void nearest_neighbor_classifier::unlearn_id(const std::string& id) {
...
+  if (labels_[label] <= 0) {
+    labels_.erase(label);
+  }
+}
+
+void nearest_neighbor_classifier::regenerate_label_counter() {
+  labels_t new_labels;
+
+  {
+    // Copy keyset of labels_ to new lanels.
```
Comment (kmaehashi) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: `lanels` => `labels`

**Comment by kmaehashi** [comment]
:+1: LGTM except for one tiny typo

**Comment by rimms** [comment]
@kmaehashi Thank you. I fixed. 💨

**Comment by kmaehashi** [comment]
😎 LGTM

**Pull request merged**